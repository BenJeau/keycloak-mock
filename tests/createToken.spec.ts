import axios from "axios";
import { setupBefore, teardownAfter, getMockInstance } from "./util";

describe("createToken", () => {
  beforeAll(setupBefore);
  afterAll(teardownAfter);

  const createInstanceAndURL = () => {
    const kmock = getMockInstance();
    const url = kmock.createURL(
      "/realms/myrealm/protocol/openid-connect/token"
    );

    return { kmock, url };
  };

  it("returns 400 without a grant_type", async () => {
    const { kmock, url } = createInstanceAndURL();

    const { status } = await axios.post(
      url,
      {
        username: "henk@gmail.com",
        password: "testPassword!",
        client_id: "test",
      },
      { validateStatus: () => true }
    );

    expect(status).toBe(400);
  });

  it("returns 400 with an invalid grant_type", async () => {
    const { kmock, url } = createInstanceAndURL();

    const { status } = await axios.post(
      url,
      {
        grant_type: "unknown",
        username: "henk@gmail.com",
        password: "testPassword!",
        client_id: "test",
      },
      { validateStatus: () => true }
    );

    expect(status).toBe(400);
  });

  it("returns 400 with missing username and password", async () => {
    const { kmock, url } = createInstanceAndURL();

    const { status } = await axios.post(
      url,
      {
        grant_type: "password",
        client_id: "test",
      },
      { validateStatus: () => true }
    );

    expect(status).toBe(400);
  });

  it("returns 400 with invalid client_id", async () => {
    const { kmock, url } = createInstanceAndURL();

    const { status } = await axios.post(
      url,
      {
        grant_type: "password",
        username: "henk@gmail.com",
        password: "testPassword!",
        client_id: "hbfsdfdf",
      },
      { validateStatus: () => true }
    );

    expect(status).toBe(400);
  });

  it("returns 403 with non existing user", async () => {
    const { kmock, url } = createInstanceAndURL();

    const { status } = await axios.post(
      url,
      {
        grant_type: "password",
        username: "bla@test.com",
        password: "testPassword!",
        client_id: "test",
      },
      { validateStatus: () => true }
    );

    expect(status).toBe(403);
  });

  it("returns 403 with wrong password", async () => {
    const { kmock, url } = createInstanceAndURL();

    const { status } = await axios.post(
      url,
      {
        grant_type: "password",
        username: "henk@gmail.com",
        password: "wrongPassword",
        client_id: "test",
      },
      { validateStatus: () => true }
    );

    expect(status).toBe(403);
  });

  it("allows a regular user to login with a password", async () => {
    const { kmock, url } = createInstanceAndURL();

    const { status, data } = await axios.post(
      url,
      {
        grant_type: "password",
        username: "henk@gmail.com",
        password: "testPassword!",
        client_id: "test",
        scope: "test",
      },
      { validateStatus: () => true }
    );

    expect(status).toBe(200);
    expect(data.access_token).toBeTruthy();
    expect(data.refresh_token).toBeTruthy();
    expect(data.token_type).toBeTruthy();
    expect(data.session_state).toBeTruthy();

    expect({
      ...data,
      access_token: null,
      refresh_token: null,
      session_state: null,
    }).toMatchSnapshot();
  });

  it("allows a client to login with a secret key", async () => {
    const { kmock, url } = createInstanceAndURL();

    const { status, data } = await axios.post(
      url,
      {
        grant_type: "client_credentials",
        client_secret: kmock.params.clientSecret,
        client_id: "test",
        scope: "test",
      },
      { validateStatus: () => true }
    );

    expect(status).toBe(200);
    expect(data.access_token).toBeTruthy();
    expect(data.refresh_token).toBeTruthy();
    expect(data.token_type).toBeTruthy();
    expect(data.session_state).toBeTruthy();

    expect({
      ...data,
      access_token: null,
      refresh_token: null,
      session_state: null,
    }).toMatchSnapshot();
  });
});
