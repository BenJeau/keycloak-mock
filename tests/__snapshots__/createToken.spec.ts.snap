// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createToken allows a client to login with a secret key 1`] = `
Object {
  "access_token": null,
  "expires_in": 3600,
  "not-before-policy": 0,
  "refresh_expires_in": 3600,
  "refresh_token": null,
  "scope": "test",
  "session_state": null,
  "token_type": "bearer",
}
`;

exports[`createToken allows a regular user to login with a password 1`] = `
Object {
  "access_token": null,
  "expires_in": 3600,
  "not-before-policy": 0,
  "refresh_expires_in": 3600,
  "refresh_token": null,
  "scope": "test",
  "session_state": null,
  "token_type": "bearer",
}
`;
